attributes.go

```diff
@@ -31,6 +32,8 @@ import (
 // kept else it will be removed. AttributeFilter must be safe to call concurrently.
 type AttributeFilter func(connect.Spec, attribute.KeyValue) bool
 
+type CustomAttributes func(context.Context, connect.AnyRequest) []attribute.KeyValue
```

Comment:
This signature is not amenable to support for streaming RPCs. I think you instead want to pass all the components of `AnyRequest` _other than_ the actual request message:
```suggestion
type CustomAttributes func(ctx context.Context, httpMethod string, requestHeaders http.Header, spec connect.Spec, peer connect.Peer) []attribute.KeyValue
```
All of these same things are available for streams, which means you can use this same callback function to get custom metrics for streaming RPCs, too. (The stream interface doesn't provide an HTTP method, but that's okay because it _always_ `http.MethodPost` for streams.)